pre_commit/languages/download.py

```diff
@@ -0,0 +1,258 @@
+from __future__ import annotations
+
+import contextlib
+import hashlib
+import os.path
+import platform
+import stat
+from base64 import standard_b64decode as b64decode
+from base64 import standard_b64encode as b64encode
+from dataclasses import dataclass
+from os import chmod
+from pathlib import Path
+from pathlib import PurePath
+from typing import AnyStr
+from typing import Generator
+from typing import IO
+from typing import Iterator
+from typing import Literal
+from typing import Sequence
+from urllib.parse import urlparse
+from urllib.request import urlopen
+
+from pre_commit import lang_base
+from pre_commit.envcontext import envcontext
+from pre_commit.envcontext import PatchesT
+from pre_commit.envcontext import Var
+from pre_commit.prefix import Prefix
+
+ENVIRONMENT_DIR = 'download'
+get_default_version = lang_base.basic_get_default_version
+run_hook = lang_base.basic_run_hook
+
+
+def get_env_patch(target_dir: str) -> PatchesT:
+    return (
+        ('PATH', (target_dir, os.pathsep, Var('PATH'))),
+    )
+
+
+@contextlib.contextmanager
+def in_env(prefix: Prefix, version: str) -> Generator[None, None, None]:
+    envdir = lang_base.environment_dir(prefix, ENVIRONMENT_DIR, version)
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+@dataclass(frozen=True)
+class Platform:
+    value: str
+
+    def __post_init__(self) -> None:
+        os, cpu = self.parts
+
+        if os not in {'linux', 'darwin', 'windows'}:
+            raise ValueError(f'invalid operating system `{os}`')
+
+        if cpu not in {'amd64', 'aarch64'}:
```

Comment:
This should match the literal values for the `cpu` property.

```suggestion
        if cpu not in {'arm64', 'arm64be', 'arm', '386', 'amd64', 'ppc', 'ppc64', 'ppc64le',}:
```

You could possibly re-use the literal values simliar to:

```py
from typing import TypeAlias, get_args

class Platform:
    Os: TypeAlias = Literal["amd64", ...]

    def __post_init__(self):
        if os not in get_args(Os):
            ...

    @property
    def os(self) -> Os:
```

Reply:
I am a bit concerned about not translating / matching all the potential returns from Python to Go.

In Go, CPUs and OSes are defined here: 
https://github.com/golang/go/blob/master/src/go/build/syslist.go

In Python, it depends on the OS implementation:
https://stackoverflow.com/questions/31107019/what-are-the-potential-results-for-platform-system
https://stackoverflow.com/questions/45124888/possible-values-for-platform-machine

Reply:
I wouldn't worry about it. Just translate the ones you are confident about: more can be added in later patches. The  important ones are `amd64`/`arm64` and `linux`/`windows`/`darwin`.

Reply:
I have decided to use a ``MappingProxyType``, to avoid copying and pasting literals. I am just getting them to return ``str`` for now. Let me know if you want me to something else differently.